src/circuit/circuit_frame.rs

```diff
@@ -5722,13 +5778,88 @@ mod tests {
             )
             .unwrap();
 
-            popcount(
+            assert!(popcount(
                 &mut cs.namespace(|| format!("popcount {}", x.to_string())),
                 &bits,
                 &alloc_popcount.hash(),
-            );
+            )
+            .is_ok());
         }
 
         assert!(cs.is_satisfied());
     }
+
+    #[test]
+    fn test_to_u32() {
+        let mut cs = TestConstraintSystem::<Fr>::new();
+        let s = &mut Store::<Fr>::default();
+        let g = GlobalAllocations::new(&mut cs.namespace(|| "global_allocations"), s).unwrap();
+        let a = s.num(42);
+        let alloc_a = AllocatedPtr::alloc_ptr(&mut cs.namespace(|| "a"), s, || Ok(&a)).unwrap();
+        let add_pow_32 = add(
+            &mut cs.namespace(|| "add pow(2, 32)"),
+            &alloc_a.hash(),
+            &g.power2_32_num,
+        )
```

Comment:
If you want to allocate a number that is 42 + 2^32, it would probably be better to just do that directly. Doing it this way (setting aside the unnecessary `AllocatedPtr`) seems to conflate a lot of things, some of which are indirectly part of what is being tested.

Reply:
Not sure I understand the point. I remember once I allocated the final result directly and the comment was to explicitly calculate the power of 2, to make it clearer. Now I am using our globals to follow the same guideline.

Reply:
I think it's two separate issues. I just meant that here the test is depending on the global being correct. Once can imagine scenarios in which the global is incorrect, but the test passes because the same incorrect value is depended on in both the test and the implementation.

Instead, if the test directly calculates the value it requires, that can't happen and the test is more targeted.

Reply:
done

Reply:
Sorry to keep picking at this, but I feel like getting a clear theory around this kind of test will be helpful. What benefit is there to computing `add_pow_32` in the circuit, rather than just computing `v` directly and using it to allocate `add_pow_32`?

Reply:
I want to show that 2^32 + 42 becomes 42 after coercion. I need to both values allocated in the circuit.

Reply:
just removed the sum from the circuit

Reply:
I thought it would be easier to demonstrate what I had in mind. I pushed a commit for consideration and discussion.

Reply:
Thanks, I see the point now, which is the isolation of what is been tested.